### ISSUE TITLE
Continuing training from pre-existing Faster R-CNN weights --> decrease in mAP

### ISSUE THREAD

**Issue by aleksispi** [endernewton/tf-faster-rcnn#311]
Hi,

I trained a ResNet-v1 Faster R-CNN detector and got the 79.8 mAP for pascal voc 2007, as expected. If I then take those weights that give 79.8 mAP and use that as initial weights for the training code, and don't run a single training iteration, but instead let it snapshot the same weights that are used in initialization, Then I get ~3 mAP lower for pascal voc 2007. Is there any explanation -- what happens in the snapshotting (or elsewhere) so that the result decrease? [I thought at first it had to to with mean-std bounding box normalization, but that does not seem to be the case.]

Best,
Aleksis

**Comment by endernewton** [comment]
Did you call the hack function that initializes the weights with the bottom layer rgb channel swapped?

Regards,
Xinlei

________________________________________

**Comment by aleksispi** [comment]
Hi Xinlei,

Thank you very much for your response. Indeed, the problem was that `self.net.fix_variables(sess, self.pretrained_model)` in the `initialize` function in `train_val.py` flips the RGB --> BGR. This is only supposed to happen given that one trains from imagenet-weights, NOT if one wants to continue training from existing Faster R-CNN weights, as I wanted. I added a flag `do_reverse:` `self.net.fix_variables(sess, self.pretrained_model, do_reverse)`, and in `resnet_v1.py` it now looks like this:

`
 def fix_variables(self, sess, pretrained_model, do_reverse=True):
    print('Fix Resnet V1 layers..')
    if do_reverse:
      print("OBS: REVERSING RGB CHANNELS -- set do_reverse=False if this"
            " is not desired!")
    else:
      print("OBS: NOT REVERSING RGB CHANNELS -- set do_reverse=True if"
            " reversing is desired!")
    with tf.variable_scope('Fix_Resnet_V1') as scope:
      with tf.device("/cpu:0"):
        # fix RGB to BGR
        conv1_rgb = tf.get_variable("conv1_rgb", [7, 7, 3, 64], trainable=False)
        restorer_fc = tf.train.Saver({self._scope + "/conv1/weights": conv1_rgb})
        restorer_fc.restore(sess, pretrained_model)
        if do_reverse:
          sess.run(tf.assign(self._variables_to_fix[self._scope + '/conv1/weights:0'], 
                             tf.reverse(conv1_rgb, [2])))
        else:
          sess.run(tf.assign(self._variables_to_fix[self._scope + '/conv1/weights:0'], 
                             conv1_rgb))
`

It now works!

Thanks again,
Aleksis

**Comment by endernewton** [comment]
glad it helped, closing.
